+toerror

ingun37 opened the pull request:
Hi! it's been a while since we talked about "BoolToError" from the Issue.
It's my first pull request to any repository ever, so hope u understand if there's any clumsy mistake 😅
Also I don't know how to handle correctly that the import path in main.go being generated pointing to my forked go package "github.com/ingun37/goderive/plugin/toerror" not "github.com/awalterschulze/goderive/plugin/toerror".
I'll be waiting for your feedback
thank you

awalterschulze commented on main.go:
```diff
@@ -53,6 +53,7 @@ import (
 	"github.com/awalterschulze/goderive/plugin/uncurry"
 	"github.com/awalterschulze/goderive/plugin/union"
 	"github.com/awalterschulze/goderive/plugin/unique"
+	"github.com/ingun37/goderive/plugin/toerror"
```
Yeah this is weird problem with Go.
What I do in this case is:
```
cd $GOPATH/src/github.com/awalterschulze
git clone https://github.com/ingun37/goderive 
```
Then your import can be 
```
github.com/ingun37/goderive/plugin/toerror
```

awalterschulze commented on plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,106 @@
+		return "", fmt.Errorf("%s, 2nd param %s, is not of type function", name, g.TypeString(typs[1]))
+	}
+	params := sig.Params()
+	if params.Len() != 1 {
```
I don't know if the function really needs an argument.
We should be able to handle any amount of arguments, especially zero

awalterschulze commented on plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,106 @@
+	if results.Len() != 2 {
+		return "", fmt.Errorf("%s, return type of the given function must be a tuple length of 2", name)
+	}
+	if !types.Identical(results.At(1).Type(), types.Typ[types.Bool]) {
```
I think you can simply do a type switch on the Type(), like so
https://github.com/awalterschulze/goderive/blob/master/derive/types.go#L64
I don't think we need to use types.Identical, but maybe you know something I don't?

ingun37 replied:
This one I didn't change because in your case (referennced code snippet) its for covering all the cases and then proceeding, in my case it's only covering one exceptional case that returns error. Either way it's petty so I'd rather not change already working code.

awalterschulze commented on plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,106 @@
+
+func (g *gen) genFuncFor(deriveFuncName string, etyp *types.Named, ftyp *types.Signature) error {
+	p := g.printer
+	// fStr := g.TypeString(ftyp) //mus be "func (a) (b, bool).."
```
remove comment

awalterschulze commented:
If this goes well and you want to contribute more.  I think we can come up with many more ideas.
Just let me know.

ingun37 commented:
I improved overall comments, made it accept functions with any parameters, added testing code for it.

ingun37 commented:
I don't think catching panics is deriveToError's job but it sounds very interesting. It's actually one of things I thought would be nice to have one using goderive! I'll think about that.

awalterschulze commented:
Great work :)

You should replace me with yourself as the Copyright of the files you created.

I would like to also solve the apparent Merge Conflict that github is reporting.
What version of go are you running?

Could you run `make travis` on your computer and update the diff?

ingun37 commented:
I'm using go1.10.4
`test/normal/derived.gen.go` in my branch has been already generated by `make` so `make travis` isn't making any difference.
Normally I would git-rebase when I don't have write permission to main branch... but if I do it's gonna make a whole new pull request right?
Or it seems very trivial conflict (typical whitespace conflict) so maybe you could do it?
Or since it's thousands of lines of auto generated code file, and if there's no other reason the file should be version-controlled, maybe git-ignore it?

ingun37 commented:
I didn't know I could resolve merge conflict myself 🙄
I followed the [github guideline](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) and now I think the conflict is gone?

ingun37 commented:
According to github [guideline]( https://help.github.com/articles/resolving-a-merge-conflict-on-github/) you can resolve such trivial conflicts like what we are having now directly from github. I think you could use it.

awalterschulze commented:
Thank you so much, this looks great :)

Oh wait I forgot the most important part.
The readme.md needs to be updated.
Please add your function to the function list.

If you want to add an example that is also great, but it is optional.
https://github.com/awalterschulze/goderive/blob/master/Contributing.md#contributing-examples

ingun37 commented:
As working on example, found a couple of bugs and fixed it.
Added example.
One thing I'm concerned is that my english isn't good enough to write formal documents so maybe some comments/documents could be misleading.

awalterschulze commented on example/plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,57 @@
+	"strings"
+)
+
+func upgradeUser(endpoint, username string) error {
```
Maybe we can come up with a better example, where we really need to a lookup in a map and then do something that might cause an error.

ingun37 replied:
replaced with [Lookup](https://golang.org/pkg/reflect/#StructTag.Lookup) method from reflect package
resolving it

awalterschulze commented on example/plugin/toerror/Readme.md:
````diff
@@ -0,0 +1,25 @@
+goderive will generate the following code:
+
+```go
+func deriveToError(err error, f func(i some_key) (a some_value, b bool)) func(i some_key) (some_value, error) {
````
Something is not right, the code generation should replace the code in the readme with the actual generated code.  Did you run `make` in the example folder?

ingun37 replied:
regenerated by make. resolving it.

awalterschulze commented:
Don't worry about your english, it looks great, I haven't noticed any lack of skill, but then again my english is also not that great :)

ingun37 commented:
As working on more examples and tests I got thinking that..
It's obvious that map-indexing and type-assertion will be the significant two in practice.
But since they are syntax not a function, you gotta interpret them into function first; that would be same old tedious repetitive codes for every types; exactly what you tried to repel by using Goderive.🤔

So I want to support following 2 more signatures to make `deriveToError` actually practical.
```go
deriveToError(error, map[A]B, A) (B, error)//map indexing.
deriveToError(error, A) (B, error)//type conversion
```

First one (map indexing) will be no problem.
But for the second one (type assertion) the returning type (B) has to be inferencable without any parameter of B... is there a history or code I can reference for such case?
Or any other opinion? Like it should be another function like..`deriveComposableIndexing` or `deriveComposableTypeAssertion`.... or let's not get too practical and let's focus and finish  deriveToError first?

ingun37 commented:
Let's move on without regarding map-indexing/type-assertion extension. They could be out of scope of `deriveToError`
I've resolved the example issues.

awalterschulze commented on example/plugin/toerror/Readme.md:
```diff
@@ -0,0 +1,35 @@
+The deriveToError function is useful for composing bool-returning functions along with traditional error-returning functions.
+
+Given the following input pseudo-code:
```
Well now it is not pseudo code, it is real code :)

awalterschulze commented on example/plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,10 @@
+)
+
+func getJsonTag(f reflect.StructTag) (string, error) {
+	return deriveToError(fmt.Errorf("json tag not exists"), f.Lookup)("json")
```
`json tag does not exist`

awalterschulze commented on example/plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,10 @@
+	"reflect"
+)
+
+func getJsonTag(f reflect.StructTag) (string, error) {
```
I love using reflect here, what a great example.
Maybe we can compose it with another function that takes a string and returns another value and error, to reflect your use case stated in the example, that this function is now easier to compose.
What do you think?

awalterschulze commented:
I have also been thinking about assertions and gets from a map and was thinking of something like:
```
deriveGet(m map[a]b, a) (b, bool)
deriveGet(c chan b) (b, bool)
```
Type assertions are much harder and I don't have an idea yet.
What I want is:
```
deriveAssert(v, float64) (float64, bool)
```
But that is impossible.
For assert it might be easier to use a library that has loads of assert functions like:
```
assertFloat64(v interface{}) (float64, bool)
```
But I don't know

ingun37 commented:
I improved `toerror` to accept any return type. (previously accepted only a tuple length of 2)
Example now demonstrates composability of `toerror` with other functions.

awalterschulze commented on example/plugin/toerror/Readme.md:
```diff
@@ -0,0 +1,55 @@
+	return deriveCompose(
+		deriveToError(fmt.Errorf("HTTP version parsing failed"), http.ParseHTTPVersion),
+		func(major, minor int) (int, error) {
+			return major, nil
```
One final nit, it would be great if this example function actually also had the possibility of return an error.
For example
`if major < x then`

awalterschulze commented on plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,127 @@
+
+// Package toerror contains the implementation of the toerror plugin, which generates the deriveToError function.
+//
+// The deriveToError function transforms return type of a function from (a, bool) into (a, error).
```
This comment needs updating to reflect variable return parameters.

awalterschulze commented on plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,127 @@
+)
+
+// NewPlugin creates a new curry plugin.
+// This function returns the plugin name, default prefix and a constructor for the curry code generator.
```
This is not the curry code generator, but the toerror code generator

awalterschulze commented on plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,127 @@
+	return derive.NewPlugin("toerror", "deriveToError", New)
+}
+
+// New is a constructor for the curry code generator.
```
same here

awalterschulze commented on plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,127 @@
+	if !types.Identical(results.At(results.Len()-1).Type(), types.Typ[types.Bool]) {
+		return "", fmt.Errorf("%s, given function must return bool as last return type. (got %v)", name, results.String())
+	}
+	return g.SetFuncName(name, errTyp, funcTyp)
```
I think we can just say
```
g.SetFuncName(name, funcTyp)
```
The error type is always going to be error, I don't expect duplicates here.

ingun37 replied:
I'll see it after 24th
I have a test😫

awalterschulze replied:
Oh yes please, tests before open source.
I can wait.

ingun37 replied:
I believe you are already aware of absence of basic type of `error` in types package.
That was the reason I carried the type all the way from parameter of `Add` to `Generate` even though I knew its just `error`
I resolved it by implementing native `types.Type` acting like basic type `error`
I think its little better this way
what do you think?

awalterschulze replied:
Very clever, I like it

awalterschulze commented on plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,127 @@
+	return types.NewVar(v.Pos(), v.Pkg(), "", v.Type())
+}
+func outs(num int, last string) string {
+	outs := make([]string, num, num)
```
`outs := make([]string, num)` is the same as `outs := make([]string, num, num)`

awalterschulze commented on plugin/toerror/toerror.go:
```diff
@@ -0,0 +1,127 @@
+	newSigType := types.NewSignature(nil, ftyp.Params(), newResultType, ftyp.Variadic())
+
+	p.P("")
+	p.P("// %s transforms sum-bool type into sum-error type. Main purpose is to make the given function composable. It returns given error when the result of the function is false.", deriveFuncName)
```
I love also think about this as a sum type, but let's try to make it more goey.

"""
%s transforms the given function's last bool type into an error type. The transformed function returns the given error when the result of the given function is false, otherwise it returns nil.
"""

awalterschulze commented on test/normal/toerror_test.go:
```diff
@@ -0,0 +1,79 @@
+
+type LocalType struct{}
+
+func true0() bool {
```
awesome :)

awalterschulze commented on test/normal/toerror_test.go:
```diff
@@ -0,0 +1,79 @@
+func true4(a, b int) (int, int, bool) {
+	return a, b, true
+}
+func true5(lt *LocalType) (*LocalType, bool) {
```
yes please :)

awalterschulze commented on test/normal/toerror_test.go:
```diff
@@ -0,0 +1,79 @@
+	return t, true
+}
+
+// func fail
```
I don't know if the comment is necessary, but otherwise this test looks much cleaner :)

awalterschulze commented:
I think we are ready to merge :)
Are you happy?

ingun37 commented:
Sure :)

awalterschulze commented:
Thank you so much for this new function, I think this really fits in well with the mission :)

awalterschulze commented:
Let me know if you want to discuss more on `get` and `assert`.

Also I saw you did quite a bit of category theory for programmers.
I am getting stuck on universal construction.
Do you have any tips or other resources that explain this differently?

ingun37 commented:
I myself have trouble understanding overall category theory let alone universal construction 😂
But I think [Cakes, Custard and Category Theory](https://www.amazon.com/Cakes-Custard-Category-Theory-Understanding/dp/1781252874)'s `universal properties` chapter builds excellent intuition about universal property/construction. But it's bare intuition, nothing technical.

awalterschulze commented:
Thank you.  I'll try that.  I added it to my list :)
https://github.com/awalterschulze/learning/blob/master/Mathematics.md

awalterschulze commented:
I think universal construction is the key part of category theory and without it, I can only imagine the parts that I understand.

ingun37 commented:
I just saw your works on mathematics WOW 🙌
I admire your enthusiasm and devotion to learning
you seem to be a very good developer

awalterschulze commented:
You'll see that most of the list does not have a lot of check marks :(
I still have a lot to learn.
I feel that I missed something while at university and I am trying to catch up.

Current development feels like hacking, compared to the math we could be using.
Maybe one day I will learn how to develop, until then I will continue hacking for money :)

ingun37 commented:
Yeah, I feel what I develop very temporary and little, compared to timeless rigorous math. I can only keep learning :)
